### ISSUE TITLE
How could mAP@0.5 from n test folders be different from one grouping all test data ?

### ISSUE THREAD

**Issue by codename5281** [ultralytics/yolov5#2227]
## ❔Question
Hello ! I have a question related to interpreting test data. 
I am testing a model against different test datasets, as well as against a union dataset in order to check how my model behaves on specific conditions vs overall.
Though, i don't understand why my model which performs good on individual test sets (minus test set 8) doesn't anymore when facing the union dataset. Shouldn't the mAP@0.5 value of the union dataset be somehow around the average of the mAP@0.5 obtained for individual datasets ? Below the obtained scores :

test set 1 | 0.873
test set 2 | 0.862
test set 3 | 0.87
test set 4 | 0.797
test set 5 | 0.807
test set 6 | 0.8
test set 7 | 0.8
test set 8 | 0.51
test set 9 | 0.738
test set 10 | 0.787
test set 11 | 0.856
test set 12 | 0.842
test set 13 | 0.862
union of test sets 1 to 13 | 0.407

Thanks in advance,
J.

**Comment by glenn-jocher** [comment]
@codename5281 AP is averaged across **classes** to produce mAP, it is not averaged across images or groups of images.

**Comment by codename5281** [comment]
Thanks for your reply, @glenn-jocher. I trained my model from the pretrained models (i.e. including the 80 classes from coco) however  as i was interested in detecting one class of objects only i used training data and annotations for this only class, Consequently, i have only one class represented in my test datasets. Is your comment still valid in this context ?

**Comment by glenn-jocher** [comment]
@codename5281 yes. The mAP computation is done across all detections together, it is not done per image or per group of images. If you want the most accurate mAP number you should test on the widest available validation set.

**Comment by codename5281** [comment]
Got it @glenn-jocher ; using the --verbose argument on test.py i finally noticed that in one of my datasets i had some labels mistakenly belonging to another class, thus generating mAP values that were then used to compute the mean accross classes in my union dataset, and mechanically bringing it down, causing this surprising scoring. Everything good now, thanks for your help !

**Comment by glenn-jocher** [comment]
@codename5281 oh good! We made a change recently that should automatically apply --verbose for datasets smaller than 50 classes to help out a bit.

But yes mAP is tricky because you can't compute it per image and then average your results, you must compute it across all images togethor.
